storage/innobase/mtr/mtr0log.cc

```diff
@@ -799,9 +811,36 @@ bool mlog_open_and_write_index(mtr_t *mtr, const byte *rec,
     n = DICT_INDEX_SPATIAL_NODEPTR_SIZE;
   }
 
+  /* Ordinal position of an existing field can't be changed with INSTANT
+  algorithm. But when it is combined with ADD/DROP COLUMN, ordinal position
+  of a filed can be changed. This bool array of size #fields in index,
+  represents if ordinal position of an existing filed is changed. */
+
+  // Move the logic to find columns that changed order before calculating
+  // the buffer size
+  std::unique_ptr<bool[]> fields_with_changed_order = nullptr;
+  if (is_versioned) {
+    fields_with_changed_order = std::make_unique<bool[]>(n);
+
+    uint16_t phy_pos = 0;
+    for (size_t i = 0; i < n; i++) {
+      dict_field_t *field = index->get_field(i);
+      const dict_col_t *col = field->col;
+
+      if (col->is_instant_added() || col->is_instant_dropped()) {
+        continue;
+      }
+      if (col->get_phy_pos() >= phy_pos) {
+        phy_pos = col->get_phy_pos();
+      } else {
+        fields_with_changed_order[i] = true;
+      }
+    }
+  }
```

Comment:
[optimization]: in this loop we can derive that the total number of changed fields and simply pass the total versioned fields to log_index_get_size_needed().

We may ignore this for now.